### PR TITLE
Refactor registry address

### PR DESCRIPTION
**Pull request by magicalne**
In this pr, we get rid of the old `short address`. Instead, we involve `registry address`. See more in [here](https://github.com/jjyr/godwoken/blob/refactor-sudt-with-registry-address/docs/CHANGES.md)

### CODE REVIEW

**Comment by Flouse** [comment]
Please update the `chain_id` [definition](https://github.com/nervosnetwork/godwoken-polyjuice/blob/compatibility-breaking-changes/docs/EVM-compatible.md#others) in docs.

File: polyjuice-tests/src/test_cases/get_chain_id.rs [hunk 1]
```diff
@@ -106,8 +104,7 @@ fn test_get_chain_id() {
 
         /* chain_id = compatible_chain_id(u32) | creator_account_id(u32) */
```
Comment (Flouse) [inline polyjuice-tests/src/test_cases/get_chain_id.rs:105]:
```suggestion
        /* chain_id = GodwokenRollupConfig#chain_id(u64) */
```

**Review by Flouse — APPROVED** [review]

File: c/contracts.h [hunk 2]
```diff
@@ -15,41 +15,42 @@
...
+                                       const uint8_t *input_src,
                                        const size_t input_size,
-                                       uint8_t** output, size_t* output_size);
+                                       uint8_t **output, size_t *output_size);
 
-int ecrecover_required_gas(const uint8_t* input, const size_t input_size,
-                           uint64_t* gas) {
+int ecrecover_required_gas(const uint8_t *input, const size_t input_size,
+                           uint64_t *gas)
+{
```
Comment (TheWaWaR) [inline c/contracts.h:53]: Why do those kind of style changes? Are those changes by hand or tool? If by tool please enforce it in Makefile and CI config.

Comment (magicalne) [reply c/contracts.h:53]: This is from `make fmt`. Should I revert back?

Comment (TheWaWaR) [reply c/contracts.h:53]: Better change it in a separated PR, otherwise it a distraction when review the changes.

File: c/polyjuice.h [hunk 3]
```diff
@@ -788,66 +871,72 @@ int load_globals(gw_context_t* ctx, uint32_t to_id) {
...
   debug_print_int("creator_account_id", g_creator_account_id);
 
   /** read rollup_script_hash and g_sudt_id from creator account */
   memcpy(g_rollup_script_hash, creator_raw_args_seg.ptr, 32);
   memcpy(&g_sudt_id, creator_raw_args_seg.ptr + 32, sizeof(uint32_t));
   debug_print_data("g_rollup_script_hash", g_rollup_script_hash, 32);
   debug_print_int("g_sudt_id", g_sudt_id);
-  /** read g_eth_addr_reg_id from creator_script_args[36..40] */
-  memcpy(&g_eth_addr_reg_id, creator_raw_args_seg.ptr + 36, sizeof(uint32_t));
+  g_eth_addr_reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
```
Comment (Flouse) [inline c/polyjuice.h:925]: If GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID is used, then we don't need g_eth_addr_reg_id now.

File: c/sudt_contracts.h [hunk 4]
```diff
@@ -215,27 +212,20 @@ int transfer_to_any_sudt(gw_context_t* ctx, const uint8_t* code_data,
   }
 
   evmc_address from_address = *((evmc_address*)(input_src + 32 + 12));
+  gw_reg_addr_t from_addr = {0};
+  from_addr.reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
+  memcpy(from_addr.addr, from_address.bytes, ETH_ADDRESS_LEN);
```
Comment (Flouse) [inline c/sudt_contracts.h:217]: Maybe we could copy address from input_src directly?

File: c/sudt_contracts.h [hunk 5]
```diff
@@ -215,27 +212,20 @@ int transfer_to_any_sudt(gw_context_t* ctx, const uint8_t* code_data,
...
-  ret = load_script_hash_by_eth_address(ctx, from_address.bytes,
-                                        from_script_hash);
-  if (ret != 0) {
-    debug_print_int("[transfer_to_any_sudt] load from_script_hash failed",
-                    ret);
-    return ERROR_TRANSFER_TO_ANY_SUDT;
-  }
+  gw_reg_addr_t to_addr = {0};
+  to_addr.reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
+  memcpy(to_addr.addr, to_address.bytes, ETH_ADDRESS_LEN);
```
Comment (Flouse) [inline c/sudt_contracts.h:224]: Maybe we could copy address from input_src directly?

File: docs/Addition-Features.md [hunk 6]
```diff
@@ -34,7 +34,7 @@ See: [Example](../polyjuice-tests/src/test_cases/evm-contracts/RecoverAccount.so
    input:
    ======
      input[ 0..32] => sudt_id (big endian)
-     input[32..64] => address (short_address)
+     input[32..64] => address (eth_address)
```
Comment (Flouse) [inline docs/Addition-Features.md:37]:
```suggestion
     input[32+12..64] => address (eth_address)
```

File: docs/Addition-Features.md [hunk 7]
```diff
@@ -54,26 +54,26 @@ See: [Example](../solidity/erc20/SudtERC20Proxy_UserDefinedDecimals.sol)
    input:
    ======
      input[ 0..32 ] => sudt_id (big endian)
-     input[32..64 ] => from_addr (short address)
-     input[64..96 ] => to_addr (short address)
+     input[32..64 ] => from_addr (eth address)
```
Comment (Flouse) [inline docs/Addition-Features.md:57]:
```suggestion
     input[32+12..64 ] => from_addr (eth address)
```

File: docs/Addition-Features.md [hunk 8]
```diff
@@ -54,26 +54,26 @@ See: [Example](../solidity/erc20/SudtERC20Proxy_UserDefinedDecimals.sol)
...
+     input[64..96 ] => to_addr (eth address)
```
Comment (Flouse) [inline docs/Addition-Features.md:58]:
```suggestion
     input[64+12..96 ] => to_addr (eth address)
```

File: docs/EVM-compatible.md [hunk 9]
```diff
@@ -42,9 +42,7 @@ When you send an Ethereum transaction, the transaction is converted to Godwoken
 ## Others
 
 * Transaction context
-  * `chain_id` consists up of two parts: [**compatible_chain_id(u32) | [creator_account_id]()(u32)**]
-    - `compatible_chain_id` is defined in Godwoken [RollupConfig](https://github.com/nervosnetwork/godwoken/blob/v1.0.0-rc1/crates/types/schemas/godwoken.mol#L64).
-    - `creator_account` is known as [the root account of Polyjuice](https://github.com/nervosnetwork/godwoken/blob/v1.0.0-rc1/docs/life_of_a_polyjuice_transaction.md#root-account--deployment).
+  * `chain_id` is reading from `RollupConfig#chain_id`. See more detail of the change in [here](https://github.com/nervosnetwork/godwoken/blob/31a09698eb8ea308019f9abc97970d46139df76b/docs/CHANGES.md#core-data-structure).
```
Comment (Flouse) [inline docs/EVM-compatible.md:45]:
```suggestion
  * `chain_id` is defined in Godwoken [RollupConfig#chain_id](https://github.com/nervosnetwork/godwoken/blob/a099f2010b212355f5504a8d464b6b70d29640a5/crates/types/schemas/godwoken.mol#L64).
```

**Review by Flouse — COMMENTED** [review]
- https://github.com/magicalne/godwoken-polyjuice/pull/2

File: polyjuice-tests/src/test_cases/contract_call_contract.rs [hunk 10]
```diff
@@ -36,23 +36,21 @@ fn test_contract_call_contract() {
         SS_INIT_CODE,
         122000,
         0,
-        block_producer_id,
+        block_producer_id.clone(),
```
Comment (Flouse) [inline polyjuice-tests/src/test_cases/contract_call_contract.rs:39]: Rename block_producer_id to block_producer?

File: c/polyjuice.h [hunk 11]
```diff
@@ -309,22 +335,20 @@ struct evmc_tx_context get_tx_context(struct evmc_host_context* context) {
...
-  int ret = context->gw_ctx->sys_get_script_hash_by_account_id(
-    context->gw_ctx,
-    context->gw_ctx->block_info.block_producer_id,
-    coinbase_script_hash
-  );
-  if (ret != 0) {
-    debug_print_int("failed to get block_producer script_hash",
-                    context->gw_ctx->block_info.block_producer_id);
+  int ret = context->gw_ctx->sys_get_script_hash_by_registry_address(context->gw_ctx,
+                                                                     &context->gw_ctx->block_info.block_producer,
```
[the remark this answers is not in the document]
Comment (magicalne) [reply c/polyjuice.h:314]: Not sure how to read script hash by `memcpy`. Can you elaborate on that?

[the remark this answers is not in the document]
Comment (Flouse) [reply c/polyjuice.h:314]: see:
- https://github.com/magicalne/godwoken-polyjuice/pull/4

File: c/polyjuice.h [hunk 12]
```diff
@@ -514,22 +498,15 @@ evmc_uint256be get_balance(struct evmc_host_context* context,
...
-  if (ret != 0) {
-    debug_print_int("[get_balance] load_script_hash_by_eth_address failed",
-                    ret);
-    context->error_code = FATAL_POLYJUICE;
-    return balance;
-  }
+  gw_reg_addr_t addr = {0};
+  addr.reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
+  memcpy(addr.addr, address->bytes, ETH_ADDRESS_LEN);
+  addr.addr_len = ETH_ADDRESS_LEN;
```
Comment (TheWaWaR) [inline c/polyjuice.h:504]: Better use a function to init `gw_reg_addr_t`, in case missing copy some field.

File: c/polyjuice.h [hunk 13]
```diff
@@ -820,23 +785,20 @@ int load_globals(gw_context_t* ctx, uint32_t to_id) {
...
   mol_seg_t rollup_config_seg;
   rollup_config_seg.ptr = ctx->rollup_config;
   rollup_config_seg.size = ctx->rollup_config_size;
-  mol_seg_t id_u32_seg = MolReader_RollupConfig_get_compatible_chain_id(&rollup_config_seg);
-  memcpy(&g_compatible_chain_id, id_u32_seg.ptr, id_u32_seg.size);
-  debug_print_int("compatible_chain_id", g_compatible_chain_id);
+  mol_seg_t id_u64_seg = MolReader_RollupConfig_get_chain_id(&rollup_config_seg);
+  memcpy(&g_chain_id, id_u64_seg.ptr, id_u64_seg.size);
+  debug_print_int("chain_id", g_chain_id);
   debug_print_int("creator_account_id", g_creator_account_id);
```
Comment (TheWaWaR) [inline c/polyjuice.h:795]: Should we assert `g_creator_account_id == GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID`  here?

File: docs/Addition-Features.md [hunk 14]
```diff
@@ -4,7 +4,7 @@
   - Add `recover_account` to recover any supported signature
   - Add `balance_of_any_sudt` to query the balance of any sudt_id account
   - Add `transfer_to_any_sudt` to transfer value by sudt_id (Must collaborate with SudtERC20Proxy_UserDefinedDecimals.sol contract)
-  - Add `eth_addr_to_gw_short_script_hash` to convert ETH address to Godwoken short script hash of the corresponding account
+  - Add `eth_addr_to_gw_script_hash` to convert ETH address to Godwoken script hash of the corresponding account
```
Comment (TheWaWaR) [inline docs/Addition-Features.md:7]: `eth_addr_to_gw_script_hash`  already removed in `contracts.h`

File: docs/Addition-Features.md [hunk 15]
```diff
@@ -54,26 +54,26 @@ See: [Example](../solidity/erc20/SudtERC20Proxy_UserDefinedDecimals.sol)
...
+     input[64+12..96 ] => to_addr (eth address)
      input[96..128] => amount (big endian)
 
    output: []
 ```
 
 See: [Example](../solidity/erc20/SudtERC20Proxy_UserDefinedDecimals.sol)
 
-### `eth_addr_to_gw_short_script_hash` Spec
+### `eth_addr_to_gw_script_hash` Spec
```
Comment (TheWaWaR) [inline docs/Addition-Features.md:66]: Already removed.

File: c/polyjuice.h (hunk 13 above)
Comment (magicalne) [reply c/polyjuice.h:793]: I think they have different roles. Shouldn't be the same.

**Review by Flouse — APPROVED** [review]

**Comment by TheWaWaR** [comment]
LGTM, please fix the CI.

**Comment by Flouse** [comment]
> LGTM, please fix the CI.

- chore(CI): fix integration-test
   https://github.com/nervosnetwork/godwoken-polyjuice/actions/runs/2162200056
   
- https://github.com/magicalne/godwoken-polyjuice/pull/5
   @magicalne pls review

**Review by TheWaWaR — APPROVED** [review]

**Pull request merged**